feat: add `blas/ext/base/dediff` (#10619)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/dediff/README.md

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_dediff( 5, x, 1, 1, p, 1, 1, a, 1, out, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double` input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **N1**: `[in] CBLAS_INT` number of indexed elements to prepend.
-   **Prepend**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to prepend after computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **N2**: `[in] CBLAS_INT` number of indexed elements to append.
-   **Append**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to append after computing differences.
-   **strideA**: `[in] CBLAS_INT` strides length for `Append`.
-   **Out**: `[out] double` output [`Float64Array`][@stdlib/array/float64]. Must have `N + N1 + N2 - 1` indexed elements.
-   **strideOut**: `[in] CBLAS_INT` stride length for `Out`.

```c

> void stdlib_strided_dediff( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT N1, const double *Prepend,  const CBLAS_INT strideP, const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, double *Out,  const CBLAS_INT strideOut );

Reviewer: ```suggestion
void stdlib_strided_dediff( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT N1, const double *Prepend, const CBLAS_INT strideP, const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, double *Out, const CBLAS_INT strideOut );
```

```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double` input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
-   **N1**: `[in] CBLAS_INT` number of indexed elements to prepend.
-   **Prepend**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to prepend after computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **offsetP**: `[in] CBLAS_INT` starting index for `Prepend`.
-   **N2**: `[in] CBLAS_INT` number of indexed elements to append.
-   **Append**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to append after computing differences.
-   **strideA**: `[in] CBLAS_INT` strides length for `Append`.
-   **offsetA**: `[in] CBLAS_INT` starting index for `Append`.
-   **Out**: `[out] double` output [`Float64Array`][@stdlib/array/float64]. Must have `N + N1 + N2 - 1` indexed elements.
-   **strideOut**: `[in] CBLAS_INT` stride length for `Out`.
-   **offsetOut**: `[in] CBLAS_INT` starting index for `Out`.

```c

> void stdlib_strided_dediff_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, const CBLAS_INT N1, const double *Prepend,  const CBLAS_INT strideP, const CBLAS_INT offsetP,  const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, const CBLAS_INT offsetA, double *Out,  const CBLAS_INT strideOut, const CBLAS_INT offsetOut );

Reviewer: ```suggestion
void stdlib_strided_dediff_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, const CBLAS_INT N1, const double *Prepend, const CBLAS_INT strideP, const CBLAS_INT offsetP, const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, const CBLAS_INT offsetA, double *Out, const CBLAS_INT strideOut, const CBLAS_INT offsetOut );
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 6 );

dediff( x.length, x, 1, 1, p, 1, 1, a, 1, out, 1 );

console.log( out );
// out => <Float64Array>[ 1.0, 2.0, 2.0, 2.0, 2.0, 11.0 ]
```

The function has the following parameters:

-   **N**: number of indexed elements.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length for `x`.
-   **N1**: number of indexed elements to `prepend`.
-   **prepend**: a [`Float64Array`][@stdlib/array/float64] containing values to prepend after computing differences.
-   **strideP**: stride length for `prepend`.
-   **N2**: number of indexed elements to `append`.
-   **append**: a [`Float64Array`][@stdlib/array/float64] containing values to append after computing differences..
-   **strideA**: strides length for `append`.
-   **out**: output [`Float64Array`][@stdlib/array/float64]. Must have `N + N1 + N2 - 1` elements.
-   **strideOut**: stride length for `out`.

> The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to differences of every other element:

Reviewer: ```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the difference between every other element:
```

console.log( 'Output:', out );
```

</section>

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dediff.h"
```

<!-- lint disable maximum-heading-length -->

#### stdlib_strided_dediff( N, \*X, strideX, N1, \*Prepend, strideP, N2, \*Append, strideA, \*Out, strideOut )

Calculates the differences between consecutive elements of a double-precision floating-point strided array.

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_dediff( 5, x, 1, 1, p, 1, 1, a, 1, out, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double` input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.

> -   **N1**: `[in] CBLAS_INT` number of indexed elements to `Prepend`.

Reviewer: ```suggestion
-   **N1**: `[in] CBLAS_INT` number of indexed elements to prepend.
```

-   **Prepend**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to prepend after computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.

> -   **N2**: `[in] CBLAS_INT` number of indexed elements to `Append`.

Reviewer: ```suggestion
-   **N2**: `[in] CBLAS_INT` number of indexed elements to append.
```

```c
void stdlib_strided_dediff( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT N1, const double *Prepend,  const CBLAS_INT strideP, const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, double *Out,  const CBLAS_INT strideOut );
```

<!-- lint disable maximum-heading-length -->

#### stdlib_strided_dediff_ndarray( N, \*X, strideX, offsetX, N1, \*Prepend, strideP, offsetP, N2, \*Append, strideA, offsetA, \*Out, strideOut, offsetOut )

Calculates the differences between consecutive elements of a double-precision floating-point strided array using alternative indexing semantics.

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_dediff_ndarray( 5, 1, x, 1, 0, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **X**: `[in] double` input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.

> -   **N1**: `[in] CBLAS_INT` number of indexed elements to `Prepend`.

Reviewer: ```suggestion
-   **N1**: `[in] CBLAS_INT` number of indexed elements to prepend.
```

-   **Prepend**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to prepend after computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **offsetP**: `[in] CBLAS_INT` starting index for `Prepend`.

> -   **N2**: `[in] CBLAS_INT` number of indexed elements to `Append`.

Reviewer: ```suggestion
-   **N2**: `[in] CBLAS_INT` number of indexed elements to append.
```

It is odd mixing capitalization here.
